keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it. 
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```golang
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx)
+  logger.V(6).Info("Updating Pod status")
+  [...]
+  logger.V(4).Info("Pod status updated", "status", newStatus)
+}
+```
+
+### New logging formats
+
+Introduction of new logging API should be paired with new output formats that can properly utilize them. In order to simplify log ingestion, but still maintain similar developer experience we would like to propose two formats. First one for production use and second one for component development.
+
+#### Production format
+
+We would like to propose a JSON as a default production format in Kubernetes. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backens (Elasticsearch, Stackdriver, BigQuery Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{"ts": "00:15:15.525108", "caller": "controller_utils.go:116", "msg": "Updating ready status", "core/node": {"name": "node-1"}, "core/pod": {"name": "nginx-1", "namespace": "default"}, "controller": {"status": false}}
+```
+
+#### Development format
+
+Dedicated for eyeballing by kubernetes developers. This format is expected to change with time as there is no need to maintain any guarantee for its compatibility. Currently proposed format is based on [glogr](https://github.com/go-logr/glogr). 
+
+Example:
+
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] controller “msg”=”Updating ready status” "core/node": {"name": "node-1"} "core/pod": {"name": "nginx-1", "namespace": "default"} "controller": {"status": false}
+```
+
+### Logging configuration
+
+ 
+A new flag will be introduced to all kubernetes components allowing users to select their preferred logging format and maintain backward compatibility.
+Proposed flag `--logging-format`, with 3 values:
```

Comment:
We should probably use a feature flag here. Command line flags are forever and this seems like a transitional thing.

Reply:
Only 1 out of 3 flag values would be transitional. We should not close possibility of alternative implementations supported in k8s.